modules/phase_field/src/userobjects/PolycrystalCircles.C

```diff
@@ -0,0 +1,190 @@
+  : PolycrystalUserObjectBase(parameters),
+    _columnar_3D(getParam<bool>("columnar_3D")),
+    _grain_num(0)
+{
+}
+
+void
+PolycrystalCircles::getGrainsBasedOnPoint(const Point & point,
+                                          std::vector<unsigned int> & grains) const
+{
+  unsigned int n_grains = _centerpoints.size();
+  unsigned int cnt = 0;
+  std::vector<bool> is_at_point;
+  is_at_point.resize(n_grains);
+
+  for (unsigned int i = 0; i < n_grains; ++i)
+  {
+    Real d_x = (point(0) - _centerpoints[i](0)) * (point(0) - _centerpoints[i](0));
+    Real d_y = (point(1) - _centerpoints[i](1)) * (point(1) - _centerpoints[i](1));
+    Real d_z = (point(2) - _centerpoints[i](2)) * (point(2) - _centerpoints[i](2));
+    Real distance = 0;
+
+    if (_columnar_3D)
+      distance = std::sqrt(d_x + d_y);
+    else
+      distance = std::sqrt(d_x + d_y + d_z);
+
+    if (distance < _radii[i])
+    {
+      is_at_point[i] = true;
+      ++cnt;
+    }
+    else
+      is_at_point[i] = false;
+  }
+
+  grains.resize(cnt);
+  unsigned int j = 0;
+
+  for (unsigned int i = 0; i < n_grains; ++i)
+  {
+    if (is_at_point[i] == true)
+    {
+      grains[j] = i;
+      ++j;
+    }
+  }
+}
+
+Real
+PolycrystalCircles::getVariableValue(unsigned int op_index, const Point & p) const
+{
+  /*
+  NOTE: This function is identical to the one found in PolycrystalVoronoi.
+  It would be best to pull the function from that file rather than rewriting it
+  here. However, there are two issues that taken together prevent me from doing
+  that:
+
+  1. This file builds on PolycrystalUserObjectBase rather than PolycrystalVoronoi
+  in order to avoid inputting the parameter grain_num, which is not needed here.
+  2. This is a virtual function and must be built either in this file or in the
+  file this builds on.
+
+  How can I get around these issues and inherit the function from PolycrystalVoronoi?
+  */
+  std::vector<unsigned int> grain_ids;
+  getGrainsBasedOnPoint(p, grain_ids);
+
+  unsigned int active_grain_on_op = invalid_id;
+  for (auto grain_id : grain_ids)
+    if (op_index == _grain_to_op[grain_id])
+    {
+      active_grain_on_op = grain_id;
+      break;
+    }
+
+  return active_grain_on_op != invalid_id ? 1.0 : 0.0;
+}
+
+void
+PolycrystalCircles::precomputeGrainStructure()
+{
+  bool readfromfile = getParam<bool>("read_from_file");
+  if (readfromfile)
+  {
+    FileName file_name = getParam<FileName>("file_name");
+    unsigned int header_length = getParam<unsigned int>("header_length");
+    std::vector<Real> data;
+    std::ifstream inFile;
+
+    MooseUtils::checkFileReadable(file_name);
+    inFile.open(file_name.c_str());
+
+    for (unsigned int i = 0; i < header_length; ++i)
+      inFile.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+
+    std::string line;
+    while (std::getline(inFile, line))
+      ++_grain_num;
+
+    inFile.close();
+
+    data.resize(_grain_num * 4);
+
+    inFile.open(file_name.c_str());
```

Comment:
I think I may be missing part of the design here. There are two smooth circle objects in this PR. One is deriving from the existing IC base class and reading data from a file. The other is using the new `PolycrystalUserObject`. You've got two different tests using these two object independently so all that's fine. So the question is, do we need both of them? What are they doing that's fundamentally different?

To answer your other question, the answer is yes. Typically you'd make some common file reader just a free standing utility function that's not part of the object. Then you could call it from multiple places. Alternatively you make what's called a "static method" which is a fancy word for a free function that happens to be "in" a class but can be called without creating an instance of that class. Instead you just call the function by specifying the namespace.  `SmoothCircleFromFileIC::readData()`.